python/eos/analysis.py

```diff
@@ -451,12 +451,24 @@ def sample_pmc(self, log_proposal, step_N=1000, steps=10, final_N=5000, rng=np.r
         log_target = pypmc.tools.indicator.merge_function_with_indicator(self.log_pdf, ind, -np.inf)
 
         # rescale log_proposal arguments to [-1, 1]
-        for component in log_proposal.components:
+        for n, component in enumerate(log_proposal.components):
             rescaled_mu = self._par_to_x(component.mu)
             rescaled_sigma = np.array([[
                 4 * component.sigma[i, j] / (bj[1] - bj[0]) / (bi[1] - bi[0]) for j, bj in enumerate(self.bounds)
                 ] for i, bi in enumerate(self.bounds)])
-            component.update(rescaled_mu, rescaled_sigma)
+            try:
+                component.update(rescaled_mu, rescaled_sigma)
+            except np.linalg.LinAlgError:
+                # Try to increase the diagonal by the smallest eigenvalue
+                smallest_eigenvalue = np.min(np.linalg.eigvalsh(rescaled_sigma))
```

Comment:
This can be a negative number, correct? I would add ``np.abs(smallest_eigenvalue)``.

I think adding the smallest eigenvalue is a bit too much, but it could work. It's better than aborting the PMC in any case!

Reply:
Ah there is a missing minus sign, thanks. If the matrix is not positive definite the smallest eigenvalue is negative, and it is the minimal number that we can add, no?

Reply:
Hm, I'm not sure that this statement is true. I assume here that the true eigenvalue is small but positive, and the numerically it accidentally turns negative.

However, I do not see the exact relation between the term added to the diagonal and the shift applied to the smallst eigenvalue! Adding to the diagonal **usually** works, but I usually try something of the order of the sqrt(machine precision).

Reply:
What I mean is that if `A.v = lambda.v` then `(A-lambda.Id).v = 0`. Do we need non-zero eigenvalues?

Reply:
Or do you mean that `np.linalg.eigvalsh(rescaled_sigma)` will return only positive values?

Reply:
The Cholesky decompostion requires that A is **positive definite**, which is equivalent that all eigenvalues are strictly positive. So yes, we need non-zero eigenvalues!

Reply:
Should I subtract twice the smallest eigenvalue or the eigenvalue times `(1 + 1e-20)`?